### ISSUE TITLE
pip install fails with missing SSL module for Windows debug jobs

### ISSUE THREAD

**Issue by nuclearsandwich** [ros2/ci#77]
dirk-thomas/vcstool#42 was released to fix a regression and did so for all but one nightly job.

The Windows Debug nightly is still failing and after looking at why, it appears that the Python packages used during the jobs have not been updating since the switch to the debug interpreter was made.

```
13:14:18 Could not fetch URL https://pypi.python.org/simple/vcstool/: There was a problem confirming the ssl certificate: Can't connect to HTTPS URL because the SSL module is not available. - skipping
13:14:18 Requirement already up-to-date: vcstool in c:\python35\lib\site-packages
```

I didn't find anything that would easily let me enable or build ssl for the debug interpreter and when I attempted to bypass https by installing with the command

```
C:\> "C:\Python35\python_d.exe" -m pip install --index-url="http://pypi.python.org/simple" --trusted-host pypi.python.org -U EmPy coverage flake8 flake8-import-order mock nose pep8 pydocstyle pyflakes vcstool
```

Python still complained about HTTPS errors. I have temporarily brought the build back online by downloading the source archive for vcstool 0.1.26 and installing it with 

```
C:\>"C:\Python35\python_d.exe" -m pip install -U --no-index --find-links="file://%USERPROFILE%/Downloads/vcstool-0.1.26.tar.gz" vcstool
```

This started with build http://ci.ros2.org/view/nightly/job/nightly_win_deb/469/ Introduced by https://github.com/ros2/ci/pull/72

Potential permanent fixes:

- Find out how to successfully bypass https when accessing pypi using the Debug interpreter
- Build or install the ssl module for the debug interpreter
- Write something that checks the pypi index and source downloads the latest versions of our packages using the standard python interpreter then passes them to the debug interpreter to be installed.

As of now I've only updated vcstool and only on icecube. Now that that's building I'm going to run around and update the other windows hosts in similar fashion.

**Comment by mikaelarguedas** [comment]
> As of now I've only updated vcstool and only on windshield. Now that that's building I'm going to run around and update the other windows hosts in similar fashion.

Do you mean icecube ? http://ci.ros2.org/view/nightly/job/nightly_win_deb/492/

> I didn't find anything that would easily let me enable or build ssl for the debug interpreter

I thought the ssl module was built by debaulf in the version of Python3 Debug that we build from source. At least it looks to be part of the archive our installation page recommends to download and builde (https://www.python.org/ftp/python/3.6.0/Python-3.6.0.tgz). I think it would be valuable to try to update one of the machins to build and use this version and see if it fixes the problem. That would cover `Build or install the ssl module for the debug interpreter` and be a more long lasting fix.

**Comment by nuclearsandwich** [comment]
> Do you mean icecube ? http://ci.ros2.org/view/nightly/job/nightly_win_deb/492/

yes. Corrected.

**Comment by nuclearsandwich** [comment]
The rest of the machines are running vcstool 0.1.26 in python_d. In fact they seemed to already have it installed (checked with `python_d -mpip list`) yet they all hit the same https errors when trying to run `python_d -mpip install ...` as well.

**Comment by wjwwood** [comment]
Do the python and python_d interpreters have different site-packages? Can we just install things with the non-debug interpreter and use them in the debug one?

I could imagine issues with things installed from pip which have C extensions not getting the debug version of those, but I don't think we have many of those. Maybe `trollius` (dependency of `osrf_pycommon`) but that's not used with Python3 anyways.

**Comment by nuclearsandwich** [comment]
> Do the python and python_d interpreters have different site-packages? Can we just install things with the non-debug interpreter and use them in the debug one?

I just assumed so because the failure occurred. But thinking further on it, the other three machines had vcstool 0.1.26 already so icecube may've just gotten on unlucky and pulled the debug job before any other job allowed the updates to work.

@mikaelarguedas will any of the tests for SROS 2 require Python's SSL module in debug? That might prevent us from bypassing ssl in debug.

**Comment by mikaelarguedas** [comment]
> @mikaelarguedas will any of the tests for SROS 2 require Python's SSL module in debug? That might prevent us from bypassing ssl in debug.

TL;DR: No we shouldnt need it for this Beta.


For now the python ssl module is only used in the tooling for certificate / key generation . There will be no use of the python ssl module in the core stack, all calls to SSL will be done in the DDS implementations themselves (AFAIK they are all using the C++ implementation of openSSL). This tooling is not tested and we will use already generated certificates for secure communication tests. The generation tool being meant to be invoked "manually" so I don't see a reason to invoke it with the debug interpreter so I think we're good.

**Comment by mikaelarguedas** [comment]
@nuclearsandwich I think I just faced this failure on a recent Windows debug job:
http://ci.ros2.org/job/ci_windows/3642/

```
"C:\Python36\python_d.exe" -m pip install -U EmPy coverage flake8 flake8-blind-except flake8-builtins flake8-class-newline flake8-comprehensions flake8-deprecated flake8-docstrings flake8-import-order flake8-quotes mock nose pep8 pydocstyle pyflakes pyparsing pytest pytest-cov pytest-runner pyyaml vcstool
pip is configured with locations that require TLS/SSL, however the ssl module in Python is not available.
Could not fetch URL https://pypi.python.org/simple/empy/: There was a problem confirming the ssl certificate: Can't connect to HTTPS URL because the SSL module is not available. - skipping
Requirement already up-to-date: EmPy in c:\python36\lib\site-packages
...
```

**Comment by clalancette** [comment]
Given the age of this issue, I'm going to close this out.  If this happens again we can always open a new one.
